### PR TITLE
[AIR] Add node rank and local world size info to session

### PR DESCRIPTION
**Pull request by ilee300a**
Signed-off-by: ilee300a <ilee300@anyscale.com>

Pipeline `node_rank` and `local_world_size` information in ray training so that they are accessible via `session.get_node_rank()` and `session.get_local_world_size()` in the training loop. 

Note : Need to add tests

Related PRs:
https://github.com/ray-project/ray/pull/29812 -- need these information for resumed training and some other mosaic library functionalities.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/tests/test_backend.py [hunk 1]
```diff
@@ -53,6 +56,20 @@ def ray_4_node_4_cpu():
     cluster.shutdown()
 
 
+@pytest.fixture
+def ray_2_node_2_cpu():
```
Comment (jiaodong) [inline python/ray/train/tests/test_backend.py:60]: can these be moved to conftest.py ?  All you need is to add `deps = [":train_lib", ":conftest"]` in the corresponding pytest build https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/train/BUILD

**Review by jiaodong — COMMENTED** [review]
PR lgtm, leaving to Amog to double check on the added args to session and backend executor (ex: Do we have any places that we use *args, **kwargs that could be impacted by this ?)

File: python/ray/train/_internal/backend_executor.py [hunk 2]
```diff
@@ -272,6 +272,37 @@ def _create_local_rank_map(self) -> Dict:
...
+        node_rank_map = {}
+
+        for world_rank in range(len(self.worker_group)):
+            worker = self.worker_group.workers[world_rank]
+            node_ip = worker.metadata.node_ip
+            if node_ip not in node_ips:
+                node_ips[node_ip] = node_cnt
+                node_cnt += 1
+            node_rank_map[world_rank] = node_ips[node_ip]
+        return node_rank_map
```
Comment (amogkam) [inline python/ray/train/_internal/backend_executor.py:301]: Can we combine the three methods into a single one? So we only iterate through the worker group metadata once

File: python/ray/train/tests/test_backend.py [hunk 3]
```diff
@@ -106,6 +123,71 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TestConfig):
         pass
 
 
+class MockWorkerGroup(WorkerGroup):
+    # Mock construct metadata
+    def add_workers(self, num_workers: int):
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:128]: I don't think we need to override `add_workers` right? Just need to override `construct_metadata`

File: python/ray/train/tests/test_backend.py (hunk 1 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py:60]: This fixture is only needed for the two specific tests I have added. 
If we move this to conftest.py then shouldn't we also move other cluster setup fixtures to the conftest.py file?

File: python/ray/train/tests/test_backend.py [hunk 4]
```diff
@@ -106,6 +123,71 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TestConfig):
...
+                )
+            )
+
+        for i in range(len(new_actors)):
+            self.workers.append(
+                Worker(actor=new_actors[i], metadata=new_actor_metadata[i])
+            )
+
+
+class MockBackendExecuter(BackendExecutor):
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:155]: We also don't need to create a MockBackendExecutor ourselves- we can leverage the `patch` functionality from `unittest.mock` module for this! https://docs.python.org/3/library/unittest.mock.html.

So if we implement a `mock_construct_metadata`, then we can just do `with path(ray.train._internal.worker_group.construct_data, mock_construct_metadata): `

File: python/ray/train/tests/test_backend.py (hunk 3 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py:128]: `construct_metadata` is not a class function and is called inside the `add_workers` so I just hardcoded inside the `add_workers`

Comment (amogkam) [reply python/ray/train/tests/test_backend.py:128]: See the comment below, but we should be able to patch the `construct_metadata` function with our own mock.

File: python/ray/train/tests/test_backend.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/tests/test_backend.py:155]: There's some examples of this in `test_backend.py` that you can take a look at

File: python/ray/train/tests/test_backend.py (hunk 3 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py:126]: Done

File: python/ray/train/_internal/backend_executor.py (hunk 2 above)
Comment (ilee300a) [reply python/ray/train/_internal/backend_executor.py]: Combined now

File: python/ray/air/session.py [hunk 5]
```diff
@@ -275,6 +275,70 @@ def train_loop_per_worker():
...
+
+    .. code-block:: python
+
+        import time
+        from ray.air import session
+        from ray.air.config import ScalingConfig
+
+        def train_loop_per_worker():
+            if torch.cuda.is_available():
+                torch.cuda.set_device(session.get_local_rank())
```
Comment (amogkam) [inline python/ray/air/session.py:290]: update this code snippet to show `get_local_world_size`?

Also let's rewrite this using the doctest syntax.

```
>>> import time
...
```

File: python/ray/air/session.py [hunk 6]
```diff
@@ -275,6 +275,70 @@ def train_loop_per_worker():
...
+            ...
+
+        train_dataset = ray.data.from_items(
+            [{"x": x, "y": x + 1} for x in range(32)])
+        trainer = TensorflowTrainer(train_loop_per_worker,
+            scaling_config=ScalingConfig(num_workers=1),
+            datasets={"train": train_dataset})
+        trainer.fit()
```
Comment (amogkam) [inline python/ray/air/session.py:330]: ditto for this code snippet

File: python/ray/train/_internal/backend_executor.py [hunk 7]
```diff
@@ -254,23 +258,61 @@ def _create_local_rank_map(self) -> Dict:
...
+                1 -> 0,
+                2 -> 1,
+                3 -> 0,
+                4 -> 1
+            }
+
         """
-        rank_mapping = {}
-        ip_dict = defaultdict(int)
+        rank_mapping = {}  # map from world rank to local rank
```
Comment (amogkam) [inline python/ray/train/_internal/backend_executor.py:289]:
```suggestion
        local_rank_mapping = {}  # map from world rank to local rank
```

File: python/ray/air/session.py [hunk 8]
```diff
@@ -275,6 +275,70 @@ def train_loop_per_worker():
...
+        from ray.air.config import ScalingConfig
+
+        def train_loop_per_worker():
+            if torch.cuda.is_available():
+                torch.cuda.set_device(session.get_local_rank())
+            ...
+
+        train_dataset = ray.data.from_items(
+            [{"x": x, "y": x + 1} for x in range(32)])
+        trainer = TensorflowTrainer(train_loop_per_worker,
```
Comment (amogkam) [inline python/ray/air/session.py:327]: the train_loop_per_worker is using torch, but we are using TensorflowTrainer here

File: python/ray/train/_internal/backend_executor.py [hunk 9]
```diff
@@ -254,23 +258,61 @@ def _create_local_rank_map(self) -> Dict:
...
+
+        for world_rank in range(len(self.worker_group)):
+            worker = self.worker_group.workers[world_rank]
+            node_ip = worker.metadata.node_ip
+            local_world_size_map[world_rank] = ip_dict[node_ip]
+
+        return rank_mapping, local_world_size_map, node_rank_map
+
+    def _get_local_world_size(self) -> Dict:
+        pass
```
Comment (amogkam) [inline python/ray/train/_internal/backend_executor.py:315]: remove this function?

File: python/ray/train/tests/test_backend.py [hunk 10]
```diff
@@ -106,6 +121,35 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TestConfig):
...
+                    node_ip=str(i % 2),
+                    hostname=str(i % 2),
+                    gpu_ids=[str(i % 2)],
+                )
+            )
+
+        for i in range(len(new_actors)):
+            self.workers.append(
+                Worker(actor=new_actors[i], metadata=new_actor_metadata[i])
+            )
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:150]: I think we can simplify the amount we are overriding here and leverage super() for most of the logic:

```python
def add_workers(self, num_workers: int):
    super().add_workers(num_workers)
    for i, worker in enumerate(self.workers):
        metadata = WorkerMetadata(node_id=0,  node_ip=str(i % 2), hostname=0,  gpu_ids=[0])
        worker.metadata = metadata
```

File: python/ray/train/tests/test_backend.py [hunk 11]
```diff
@@ -106,6 +121,35 @@ def on_shutdown(self, worker_group: WorkerGroup, backend_config: TestConfig):
         pass
 
 
+class MockWorkerGroup(WorkerGroup):
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:124]: clarify in the name that this is specific for fake multi-node testing

```suggestion
class MockMultiNodeWorkerGroup(WorkerGroup):
```

File: python/ray/train/tests/test_backend.py [hunk 12]
```diff
@@ -169,6 +213,32 @@ def train_func():
...
+    config = TestConfig()
+    with patch.object(WorkerGroup, "add_workers", MockWorkerGroup.add_workers):
+        e = BackendExecutor(config, num_workers=4)
+        e.start()
+
+        def train_func():
+            return session.get_local_world_size()
+
+        e.start_training(train_func, dataset_spec=EMPTY_RAY_DATASET_SPEC)
+        assert set(e.finish_training()) == {2, 2, 2, 2}
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:226]: Btw, just to test a more complex case, I'm wondering if we should use 3 workers with expected results being `{2,1,2}`

File: python/ray/air/session.py [hunk 13]
```diff
@@ -275,6 +275,72 @@ def train_loop_per_worker():
     return session.local_rank
 
 
+@_warn_session_misuse(default_value=0)
+def get_local_world_size() -> int:
+    """Get the local rank of this worker (rank of the worker on its node).
+
+    Example:
+        >>> import time
+        >>> import os
```
Comment (amogkam) [inline python/ray/air/session.py:284]: we can remove these imports- they are not being used anymore

File: python/ray/train/tests/test_backend.py (hunk 1 above)
Comment (jiaodong) [reply python/ray/train/tests/test_backend.py:58]: we had a bad pattern that redefined a bunch of fixtures in each test file in the past that some of them are being addressed already, if this fixture is only used in this file im fine living it as-is, but please keep this context in mind when you come across other fixtures like the ones above, we might still have a few hanging around that are defined more than once.

File: python/ray/train/tests/test_backend.py (hunk 10 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py]: updated!

**Review by jiaodong — APPROVED** [review]

File: python/ray/train/tests/test_backend.py [hunk 14]
```diff
@@ -169,6 +131,32 @@ def train_func():
...
+    config = TestConfig()
+    with patch.object(WorkerGroup, "add_workers", mock_add_workers):
+        e = BackendExecutor(config, num_workers=3)
+        e.start()
+
+        def train_func():
+            return session.get_local_world_size()
+
+        e.start_training(train_func, dataset_spec=EMPTY_RAY_DATASET_SPEC)
+        assert set(e.finish_training()) == {2, 1}
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:144]: Why do we only have 2 results if num_workers is 3?

File: python/ray/train/tests/test_backend.py [hunk 15]
```diff
@@ -169,6 +131,32 @@ def train_func():
...
+    config = TestConfig()
+    with patch.object(WorkerGroup, "add_workers", mock_add_workers):
+        e = BackendExecutor(config, num_workers=3)
+        e.start()
+
+        def train_func():
+            return session.get_node_rank()
+
+        e.start_training(train_func, dataset_spec=EMPTY_RAY_DATASET_SPEC)
+        assert set(e.finish_training()) == {0, 1}
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:157]: Why is there only 2 results if num_workers is 3?

File: python/ray/train/tests/test_backend.py [hunk 16]
```diff
@@ -169,6 +131,32 @@ def train_func():
...
+        e.start()
+
+        def train_func():
+            return session.get_local_world_size()
+
+        e.start_training(train_func, dataset_spec=EMPTY_RAY_DATASET_SPEC)
+        assert set(e.finish_training()) == {2, 1}
+
+
+def test_node_ranks(ray_2_node_3_worker):
```
Comment (amogkam) [inline python/ray/train/tests/test_backend.py:147]: Btw you don't actually need to create a new fixture here- you can continue to use ray_2_node_2_cpu even if  `num_workers` is 3.

File: python/ray/train/tests/test_backend.py (hunk 15 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py:157]: it's a set

Comment (amogkam) [reply python/ray/train/tests/test_backend.py:157]: why use a set? we're not fully testing the behavior then, right?

Comment (ilee300a) [reply python/ray/train/tests/test_backend.py]: updated

File: python/ray/train/tests/test_backend.py (hunk 1 above)
Comment (ilee300a) [reply python/ray/train/tests/test_backend.py]: Fixtures have been moved to conftest and conftest is added as a BUILD dependency

File: python/ray/air/session.py [hunk 17]
```diff
@@ -275,6 +275,66 @@ def train_loop_per_worker():
...
+        ...     [{"x": x, "y": x + 1} for x in range(32)])
+        >>> trainer = TorchTrainer(train_loop_per_worker,
+        ...     scaling_config=ScalingConfig(num_workers=1),
+        ...     datasets={"train": train_dataset})
+        >>> trainer.fit() # doctest: +SKIP
+    """
+    session = _get_session()
+    if not isinstance(session, _TrainSessionImpl):
+        raise RuntimeError(
+            "`get_local_rank` can only be called for TrainSession! "
```
Comment (amogkam) [inline python/ray/air/session.py:301]:
```suggestion
            "`get_local_world_size` can only be called for TrainSession! "
```

File: python/ray/air/session.py [hunk 18]
```diff
@@ -275,6 +275,66 @@ def train_loop_per_worker():
...
```
Comment (amogkam) [inline python/ray/air/session.py:331]:
```suggestion
            "`get_node_rank` can only be called for TrainSession! "
```

**Review by amogkam — APPROVED** [review]
Thanks @ilee300a, lgtm! Just some typos in the error message. Let's fix those and please ping again once CI passes!

**Pull request merged**